docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+
+Bricks design explained
+=======================
+
+This is a runnable tutorial, which means you can copy its fragments 
+to an IPython notebook and run. The import section follows:
+
+    >>> import theano
+    >>> theano.config.compute_test_value = 'raise'
+    >>> from theano import tensor
+
+Please keep in mind that the code in this tutorial was written for illustration
+purposes only. Often only the common case is handled and corner cases are
+ignored.
+
+Introduction
+------------
+
+Deep learning (DL) models implementing using Theano share numerous
+building blocks which become more and more standartized over time.
+Factoring out such components is a very lucrative idea, and many people
+have fallen victims to the temptation to develop a library of DL
+building blocks. The list of rather well-known projects includes
+`PyLearn2`_, `Groundhog`_, `Lasagne`_, `Keras`_ and of course Blocks. Many people and
+research groups have smaller and more specialized frameworks.
+
+This document contains speculations on how such libraries should be
+designed. We will start with the simplest approach possible and show
+step by step what benefits more advanced methods bring. The final goal is
+to explain the ideas at the core of the respective part of Blocks. 
+In Blocks the components assisting to build the computation
+graph are called "bricks". Bricks are a unified representation for
+components of all levels: basic ones, like activation functions, and
+complex ones, like attention mechanism and a generic sequence generator.
+Reading this document should answer many questions that people
+often ask about bricks, such as "why so many decorators", "why is not my
+brick initialized in the constructor" and others, which are asked not so
+often, but still daunt people' minds.
+
+.. _PyLearn2: https://github.com/lisa-lab/pylearn2
+.. _Groundhog: https://github.com/lisa-groundhog/GroundHog
+.. _Lasagne: https://github.com/Lasagne/Lasagne
+.. _Keras: https://github.com/fchollet/keras
+
+Design for basic components
+---------------------------
+
+The toy task
+~~~~~~~~~~~~
+
+To make our discussion less abstract, let us consider an example. D,
+a novice in deep learning, has learnt about fully-connected feed-forward
+and recurrent networks (FFN and RNN respectively). He plans to use them
```

Comment:
I've never seen FFN or FNN used much at all, and would just avoid them altogether.